src/nn/unbiased_linear.rs

```diff
@@ -0,0 +1,286 @@
+/// between [-1 / sqrt(I), 1 / sqrt(I)].
+///
+/// # Generics
+/// - `I` The "input" size of vectors & matrices.
+/// - `O` The "output" size of vectors & matrices.
+///
+/// # Examples
+/// `UnbiasedLinear<5, 2>` can act on vectors with 5 elements, and results in vectors with 2 elements.
+/// ```rust
+/// # use dfdx::prelude::*;
+/// # let dev: Cpu = Default::default();
+/// type Model = UnbiasedLinear<5, 2>;
+/// let model = dev.build_module::<Model, f32>();
+/// // single item forward
+/// let _: Tensor<Rank1<2>, f32, _> = model.forward(dev.zeros::<Rank1<5>>());
+/// // batched forward
+/// let _: Tensor<Rank2<10, 2>, f32, _> = model.forward(dev.zeros::<Rank2<10, 5>>());
+/// ```
+#[derive(Debug, Clone)]
+pub struct UnbiasedLinear<const I: usize, const O: usize, E: Dtype, D: DeviceStorage> {
+    /// Transposed weight matrix, shape (I, O)
+    pub weight: Tensor<Rank2<O, I>, E, D>,
+}
+
+impl<const I: usize, const O: usize, E: Dtype, D: DeviceStorage> NonMutableModule
+    for UnbiasedLinear<I, O, E, D>
+{
+}
+
+impl<const I: usize, const O: usize, E: Dtype + Float + SampleUniform, D: Device<E>>
+    BuildModule<D, E> for UnbiasedLinear<I, O, E, D>
+{
+    fn try_build(device: &D) -> Result<Self, D::Err> {
+        let b: E = E::ONE / E::from_usize(I).unwrap().sqrt();
+        let weight = device.try_sample(Uniform::new(-b, b))?;
+        Ok(Self { weight })
+    }
+}
+
+impl<const I: usize, const O: usize, E: Dtype + Float + SampleUniform, D: SampleTensor<E>>
+    TensorCollection<E, D> for UnbiasedLinear<I, O, E, D>
+{
+    fn iter_tensors<V: ModuleVisitor<Self, E, D>>(visitor: &mut V) -> Result<(), V::Err> {
+        visitor.visit_tensor(
+            "weight",
+            |s| &s.weight,
+            |s| &mut s.weight,
+            TensorOptions::reset_with(|t| {
+                let b: E = E::ONE / E::from_usize(I).unwrap().sqrt();
+                t.try_fill_with_distr(Uniform::new(-b, b))
+            }),
+        )
+    }
+}
+
+impl<const I: usize, const O: usize, E: Dtype, D1: Device<E>, D2: Device<E>> ToDevice<D2>
+    for UnbiasedLinear<I, O, E, D1>
+{
+    type Output = UnbiasedLinear<I, O, E, D2>;
+    fn to_device(&self, device: &D2) -> Self::Output {
+        UnbiasedLinear {
+            weight: self.weight.to_device(device),
+        }
+    }
+}
+
+impl<const I: usize, const O: usize, E: Dtype, D: Device<E>, T> Module<T>
+    for UnbiasedLinear<I, O, E, D>
+where
+    T: SplitTape + TryMatMul<Tensor<Rank2<I, O>, E, D, T::Tape>> + HasErr<Err = D::Err>,
+    T::Tape: Tape<D>,
+    for<'a> Bias1D<'a, O, E, D>: Module<T::Output, Output = T::Output, Error = D::Err>,
+{
+    type Output = T::Output;
+    type Error = D::Err;
+
+    /// 1d forward using [matmul()] and [add()].
+    fn try_forward(&self, x: T) -> Result<Self::Output, D::Err> {
+        x.try_matmul(self.weight.retaped::<T::Tape>().try_permute()?)
+    }
+}
+
+#[derive(Clone, Debug)]
+struct Bias1D<'a, const M: usize, E: Dtype, D: DeviceStorage> {
```

Comment:
You can remove this since this doesn't have bias 😀

Reply:
Done !